---
title: "Introduction to intCC"
author: "Can Huang"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    theme: united
vignette: >
  %\VignetteIndexEntry{R package intCC}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Installation

To install and load intCC

```{r,  eval=FALSE,warning=FALSE, cache=TRUE}
# install.packages("devtools")
devtools::install_github("candsj/intCC")
```

```{r, eval=FALSE,warning=FALSE, cache=TRUE}
library("intCC")
```


# Introduction
Analysis of omic datasets has become more and more important recently. It could help to define cancer subtypes and reveal novel discoveries. How to integrate omic datasets raised much-growing attention. The Cancer Genome Atlas Program began in 2006 and generated over 2.5 petabytes of genomic, epigenomic, transcriptomic, and proteomic data. This greatly improved the development of omic datasets research. Various integrative clustering methods and tools have been proposed. We proposed two-layer intCC method to deal with some limitations existed.

# Usage of intCC

The R package **intCC** (**i**ntegrative **C**onsensus **C**lustering) introduces intCC functions inspired by consensus clustering methods.

The main function in this package is `intCC`. Users input a list of datasets, clustering methods, the number of clutsers for separate datasets and global, criterions and so on. Then consensus clustering will be applied to each dataset to derive consensus matrix. Depends on 1 layer or 2 layer, consensus matrix will be weighted and combined for each dataset and final clustering. Finally user could choose pam or hierarchical clustering method to derive the clustering results. Some useful functions are also included in this package: 

* `criterion`. This function takes a list of consensus matrix, a list of corresponding clustering label results and a vector of corresponding number of clusters. It could use Silhouette, widestGap, dunns and dunn2s index as criterion to choose the best number of clusters.

* `weightcal`. This function takes a list of consensus matrix and clustering label results. Then weights are calculated based on ratio of in-cluster proportion to out-of-cluster proportion using the cluster estimated by the algorithm itself. It could help to estimate weights for different methods of each methods or weights for consensus matrix of each dataset.

* `plot_CM`. This function takes a list of consensus matrix and clustering label results and dataset if provided. Then heatmap could be generated for consensus matrix and dataset if provided.

The other function needed for this is `consensusCluster`. This function can be found in the R package `coca` and is used to perform consensus clustering on one dataset and obtain a co-clustering matrix  (Monti et al. 2003). We made some changes to this function to allow users to define more distance metric and sample percentage for features.


# How intCC works 

First, we load four different type datasets with the same clustering structure (3 clusters of equal size) and different levels of noise.

```{r load data, fig.show='hold', eval=FALSE,warning=FALSE, cache=TRUE}
  load(system.file("extdata", "exampleData.RData", package = "intCC"))
  normData=exampleData[[1]]
  binomData=exampleData[[2]]
  poisData=exampleData[[3]]
  multData=exampleData[[4]]
  lopoisData=log(poisData+1)
```

Suppoese we only want to apply one method like k-means to each dataset. We could use the `consensusclustering` function to compute a consensus matrix for each dataset. The function will return a list of consensus matrix and clustering result.

```{r consensus_cluster1, fig.show='hold', eval=FALSE,warning=FALSE, cache=TRUE}
  normRes=consensuscluster(normData,K=3,B=1000,pItem = 0.8,pFeature = 0.8 ,clMethod ="kmeans",finalclmethod="pam")
  binomRes=consensuscluster(binomData,K=3,B=1000,pItem = 0.8,pFeature = 0.8 ,clMethod ="kmeans",finalclmethod="pam")
  poisRes=consensuscluster(lopoisData,K=3,B=1000,pItem = 0.8,pFeature = 0.8 ,clMethod ="kmeans",finalclmethod="pam")
  multRes=consensuscluster(multData,K=3,B=1000,pItem = 0.8,pFeature = 0.8 ,clMethod ="kmeans",finalclmethod="pam")
```

Then we could perform 1 layer intCC. We combine consensus clustering results from 4 datasets and calculate the weight for each dataset. Then we calculate the weighted consensus matrix and apply pam method to derive the final clustering result.

```{r 1 layer intCC1, fig.show='hold', eval=FALSE,warning=FALSE, cache=TRUE}
  #combine consensus clustering results 
  res_1layer <- c(list(normRes),list(binomRes),list(poisRes),list(multRes))

  #calculate weight and weighted consensus matrix
  weights <-weightcal(res.1layer)
  wcm=weights[1]*res.1layer[[1]]$consensusMatrix+weights[2]*res.1layer[[2]]$consensusMatrix+
    weights[3]*res.1layer[[3]]$consensusMatrix+weights[4]*res.1layer[[4]]$consensusMatrix
  
  #apply pam method to derive the final clustering result
  distances <- stats::as.dist(1 - wcm)
  weight_clusterLabels <- pam(distances,3,diss = TRUE,metric = "euclidean" )$clustering
  ari.weight<- adjustedRandIndex(true.class, weight_clusterLabels)
```

The same can be done simply using the function `intCC`. The function returns clustering results, weights and weighted consensus matrix.

```{r 1 layer intCC2, fig.show='hold', warning=FALSE, eval=FALSE,cache=TRUE}
#1 layer
  res.1layer <- intCC(exampleData, method="1 layer", individualK = rep(3, 4),
                                   globalK = 3, pFeature = 0.8 ,ccClMethods = "kmeans",
                                    ccDistHCs = "euclidean",hclustMethod = "average",finalclmethod="hclust",
                                    finalhclustMethod = "average",Silhouette=TRUE)
```

Suppose we don't know which clustering method is better and want to apply two methods like k-means and hierarchical clustering to each dataset. As we already finished the consensus clustering using k-means method, we do the consensus clustering using hierarchical clustering first.

```{r consensus_cluster2, fig.show='hold', eval=FALSE,warning=FALSE, cache=TRUE}

  normRes1=consensuscluster(normData,K=3,B=1000,pItem = 0.8,pFeature = 0.8 ,clMethod ="hclust",finalclmethod="pam")
  binomRes1=consensuscluster(binomData,K=3,B=1000,pItem = 0.8,pFeature = 0.8 ,clMethod ="hclust",dist = "binary",finalclmethod="pam")
  poisRes1=consensuscluster(poisData,K=3,B=1000,pItem = 0.8,pFeature = 0.8 ,clMethod ="hclust",dist = "jaccard",finalclmethod="pam")
  multRes1=consensuscluster(multData,K=3,B=1000,pItem = 0.8,pFeature = 0.8 ,clMethod ="hclust",dist = "hamming",finalclmethod="pam")
```  

Then we could perform 2 layer intCC. We combine k-means consensus clustering result and hierarchical clustering consensus clustering result for each dataset first and then apply pam method to each weighted consensus matrix.  

```{r 2 layer intCC1, fig.show='hold', eval=FALSE,warning=FALSE, cache=TRUE}  
  #norm
  #combine results using 2 methods
  res.norm <- c(list(normRes),list(normRes1))
  #calculate the weight and weighted consensus matrix
  weight.norm<- weightcal(res.norm)
  wcm=weight.norm[1]*res.norm[[1]]$consensusMatrix+weight.norm[2]*res.norm[[2]]$consensusMatrix
  #apply pam method to weighted consensus matrix
  distances <- stats::as.dist(1 - wcm)
  weight_clusterLabels <-pam(distances,3,diss = TRUE,metric = "euclidean" )$clustering
  weightnormRes=list(consensusMatrix=wcm,class=weight_clusterLabels)
  
  #binom
  #combine results using 2 methods
  res.binom <- c(list(binomRes),list(binomRes1))
  #calculate the weight and weighted consensus matrix
  weight.binom <- weightcal(res.binom)
  wcm=weight.binom[1]*res.binom[[1]]$consensusMatrix+weight.binom[2]*res.binom[[2]]$consensusMatrix
  #apply pam method to weighted consensus matrix
  distances <- stats::as.dist(1 - wcm)
  weight_clusterLabels <- pam(distances,3,diss = TRUE,metric = "euclidean" )$clustering
  weightbinomRes=list(consensusMatrix=wcm,class=weight_clusterLabels)
  
  #poisson
  #combine results using 2 methods
  res.pois <- c(list(poisRes),list(poisRes1))
  #calculate the weight and weighted consensus matrix
  weight.pois <- weightcal(res.pois)
  wcm=weight.pois[1]*res.pois[[1]]$consensusMatrix+weight.pois[2]*res.pois[[2]]$consensusMatrix
  #apply pam method to weighted consensus matrix
  distances <- stats::as.dist(1 - wcm)
  weight_clusterLabels <-pam(distances,3,diss = TRUE,metric = "euclidean" )$clustering
  weightpoisRes=list(consensusMatrix=wcm,class=weight_clusterLabels)
  
  #mult
  #combine results using 2 methods
  res.mult <- c(list(multRes),list(multRes1))
  #calculate the weight and weighted consensus matrix
  weight.mult <- weightcal(res.mult)
  wcm=weight.mult[1]*res.mult[[1]]$consensusMatrix+weight.mult[2]*res.mult[[2]]$consensusMatrix
  #apply pam method to weighted consensus matrix
  distances <- stats::as.dist(1 - wcm)
  weight_clusterLabels <- pam(distances,3,diss = TRUE,metric = "euclidean" )$clustering
  weightmultRes=list(consensusMatrix=wcm,class=weight_clusterLabels)
```

We combine weighted consensus clustering results from 4 datasets and calculate the weight for each dataset. Then we calculate the weighted of weighted consensus matrix and apply pam method to derive the final clustering result.

```{r 2 layer intCC2, fig.show='hold', eval=FALSE,warning=FALSE, cache=TRUE}    
  #2 layer intCC
  #combine weighted consensus clustering results
  res.2layer <- c(list(weightnormRes),list(weightbinomRes),list(weightpoisRes),list(weightmultRes))

  #calculate weight and the weighted of weighted consensus matrix
  weight.wcm <- weightcal(res.2layer)
  wcm=weight.wcm[1]*res.2layer[[1]]$consensusMatrix+weight.wcm[2]*res.2layer[[2]]$consensusMatrix+
    weight.wcm[3]*res.2layer[[3]]$consensusMatrix+weight.wcm[4]*res.2layer[[4]]$consensusMatrix
  
  #apply pam method to derive the final clustering result
  distances <- stats::as.dist(1 - wcm)
  weight_clusterLabels <- pam(distances,3,diss = TRUE,metric = "euclidean" )$clustering
  ari.wcm <- adjustedRandIndex(true.class, weight_clusterLabels)
```

The same can be done simply using the function `intCC`.

```{r 2 layer intCC3, fig.show='hold', warning=FALSE, eval=FALSE,cache=TRUE}
  #2 layer
  data_for_intCC <- list()
  data_for_intCC[[1]]=normData
  data_for_intCC[[2]]=binomData
  data_for_intCC[[3]]=lopoisData
  data_for_intCC[[4]]=multData
  data_for_intCC[[5]]=normData
  data_for_intCC[[6]]=binomData
  data_for_intCC[[7]]=poisData
  data_for_intCC[[8]]=multData
  res.2layer <- intCC(data_for_intCC, method="2 layer", individualK = rep(3, 8),
                     globalK = 3, pFeature = 0.8 ,ccClMethods = c("kmeans","kmeans","kmeans","kmeans",
                                                                  "hclust","hclust","hclust","hclust"),
                     ccDistHCs = c("euclidean","euclidean","euclidean","euclidean",
                                   "euclidean","binary","jaccard","hamming"),hclustMethod =                                 "average",finalclmethod="hclust",finalhclustMethod = "average",Silhouette=TRUE)
```

Heatmap could be generated by `plot_CM` function. `plot_CM` function will return a heatmap of consensus matrix if only consensus clustering result provided.

```{r heatmap1, fig.show='hold', warning=FALSE, eval=FALSE,cache=TRUE}
#Heatmap for consensus matrix
plot_CM(result=normRes)
```

`plot_CM` function will return a heatmap of consensus matrix and a heatmap of dataset if both original dataset and consensus clustering result provided.

```{r heatmap2, fig.show='hold', warning=FALSE, eval=FALSE,cache=TRUE}
#Heatmap for consensus matrix and dataset if dataset provided
plot_CM(data=normData,result=normRes)
```

# Session info

```{r}
sessionInfo()
```


# References 

C. Huang and P. Kuan. (2023). intCC: An efficient weighted integrative consensus clustering of multimodal data. Submitted.
A. Cabassi and P. D. Kirk. (2020). Multiple kernel learning for integrative consensus clustering of omic datasets. Bioinformatics 36, 4789.


